File: amazon/Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s==null || s.length() < 1) return "";
        // int start = 0, end = 0;
        String res = "";
        for (int i = 0; i < s.length(); i++){
            int l1 = helper(s, i, i);
            int l2 = helper(s, i, i+1);
            int l = Math.max(l1, l2);
            if (l > res.length())
                res = s.substring(i - (l - 1) / 2, i + l / 2+1);
        }
        return res;
    }
    
    public int helper(String s, int left, int right){
        int L = left, R = right;
        while (L >= 0 && R < s.length() && s.charAt(L) == s.charAt(R)) {
            L--;
            R++;
        }
        return R - L - 1;
    }


}

File: amazon/longest_palindromic_substring.cpp
//https: //leetcode.com/problems/longest-palindromic-substring/submissions/
#include <vector>
#include <unordered_map>
#include <iostream>
#include <climits>
#include <string>
#include <queue>
using namespace std;

class Solution
{
public:
  string longestPalindrome(string s)
  {
    if (s.size() == 0)
      return "";
    int mid = 0, left = 0, right = 0;
    int max_len = 0;
    int max_start = 0;
    while (mid < s.size())
    {
      left = right = mid;
      while (right + 1 < s.size() && s[right] == s[right + 1])
        right++;
      mid = right + 1;
      while (left - 1 >= 0 && right + 1 < s.size() && s[left - 1] == s[right + 1])
      {
        left--;
        right++;
      }
      if (right - left + 1 > max_len)
      {
        max_len = right - left + 1;
        max_start = left;
      }
    }
    return s.substr(max_start, max_len);
  }
};